starknet_py/net/gateway_client.py

```diff
@@ -0,0 +1,264 @@
+                f"No contract found with following identifier {block_identifier}"
+            )
+
+        return ContractCodeSchema().load(res, unknown=EXCLUDE)
+
+    async def estimate_fee(self, tx: InvokeFunction) -> int:
+        res = await self._feeder_gateway_client.post(
+            method_name="estimate_fee", payload=InvokeFunction.Schema().dump(tx)
+        )
+        # TODO should we have better type validation here?
+        return res["amount"]
+
+    async def call_contract(
+        self,
+        invoke_tx: InvokeFunction,
+        block_hash: Optional[Hash] = None,
+        block_number: Optional[int] = None,
+    ) -> List[int]:
+        block_identifier = GatewayClient._get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+
+        res = await self._feeder_gateway_client.post(
+            method_name="call_contract",
+            params=block_identifier,
+            payload=InvokeFunction.Schema().dump(invoke_tx),
+        )
+
+        # TODO should we have better type validation here?
+        return [int(v, 16) for v in res["result"]]
+
+    async def add_transaction(self, tx: StarknetTransaction) -> SentTransaction:
+        res = await self._gateway_client.post(
+            method_name="add_transaction",
+            payload=StarknetTransaction.Schema().dump(obj=tx),
+        )
+        return SentTransactionSchema().load(res, unknown=EXCLUDE)
+
+    async def deploy(
+        self,
+        contract: Union[ContractDefinition, str],
+        constructor_calldata: List[int],
+        salt: Optional[int] = None,
+    ) -> SentTransaction:
+        if isinstance(contract, str):
+            contract = ContractDefinition.loads(contract)
+
+        res = await self.add_transaction(
+            tx=Deploy(
+                contract_address_salt=ContractAddressSalt.get_random_value()
+                if salt is None
+                else salt,
+                contract_definition=contract,
+                constructor_calldata=constructor_calldata,
+            )
+        )
+
+        # TODO maybe improve/remove this exception?
+        if res.code != StarkErrorCode.TRANSACTION_RECEIVED.name:
+            raise Exception("Transaction not received")
+
+        return res
+
+    @staticmethod
+    def _get_block_identifier(
+        block_hash: Optional[Hash] = None, block_number: Optional[int] = None
+    ) -> dict:
+        if block_hash is not None and block_number is not None:
+            raise ValueError(
+                "Block_hash and block_number parameters are mutually exclusive."
+            )
+
+        # TODO gateway now supports latest block
+        if block_hash == "latest":
+            block_hash = "pending"
+
+        if block_hash is not None:
+            return {"blockHash": convert_to_felt(block_hash)}
+
+        if block_number is not None:
+            return {"blockNumber": block_number}
+
+        return {}
+
+
+# TODO rename
+class StarknetClient:
```

Comment:
I though it'd make more sense to separate actual rpc/rest api clients from our higher level `Client`s implementations. It could be renamed but I don't think it'd be good to make it a part of `Client`s itself.

Reply:
What's the potential use case for this indirection?

Reply:
Renamed to <>HttpClient